src/main/java/org/apache/maven/shared/artifact/filter/PatternIncludesArtifactFilter.java

```diff
@@ -107,322 +147,831 @@ public boolean include( final Artifact artifact )
      */
     protected boolean patternMatches( final Artifact artifact )
     {
-        return positiveMatch( artifact ) == Boolean.TRUE || negativeMatch( artifact ) == Boolean.FALSE;
+        // Check if the main artifact matches
+        char[][] artifactGatvCharArray = new char[][] {
+                            emptyOrChars( artifact.getGroupId() ),
+                            emptyOrChars( artifact.getArtifactId() ),
+                            emptyOrChars( artifact.getType() ),
+                            emptyOrChars( artifact.getBaseVersion() )
+                    };
+        Boolean match = match( artifactGatvCharArray );
+        if ( match != null )
+        {
+            return match;
+        }
+
+        if ( actTransitively )
+        {
+            final List<String> depTrail = artifact.getDependencyTrail();
+
+            if ( depTrail != null && depTrail.size() > 1 )
+            {
+                for ( String trailItem : depTrail )
+                {
+                    char[][] depGatvCharArray = tokenizeAndSplit( trailItem );
+                    match = match( depGatvCharArray );
+                    if ( match != null )
+                    {
+                        return match;
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private Boolean match( char[][] gatvCharArray )
+    {
+        if ( simplePatterns != null && simplePatterns.size() > 0 )
+        {
+            // We add the parts one by one to the builder
+            StringBuilder sb = new StringBuilder();
+            for ( int i = 0; i < 4; i++ )
+            {
+                if ( i > 0 )
+                {
+                    sb.append( ":" );
+                }
+                sb.append( gatvCharArray[i] );
+                Map<String, Pattern> map = simplePatterns.get( i );
+                if ( map != null )
+                {
+                    // Check if one of the pattern matches
+                    Pattern p = map.get( sb.toString() );
+                    if ( p != null )
+                    {
+                        patternsTriggered.add( p );
+                        return true;
+                    }
+                }
+            }
+        }
+        // Check all other patterns
+        for ( Pattern pattern : patterns )
+        {
+            if ( pattern.matches( gatvCharArray ) )
+            {
+                patternsTriggered.add( pattern );
+                return !( pattern instanceof NegativePattern );
+            }
+        }
+
+        return null;
     }
 
     /**
-     * @param artifactId add artifact to the filtered artifacts list.
+     * @param artifact add artifact to the filtered artifacts list.
      */
-    protected void addFilteredArtifactId( final String artifactId )
+    protected void addFilteredArtifactId( final Artifact artifact )
```

Comment:
The  method name is misleading because it does not operate on ids anymore.

Reply:
Fixed